src/sinks/datadog/logs/sink.rs

```diff
@@ -158,96 +128,134 @@ impl From<io::Error> for RequestBuildError {
     }
 }
 
+impl From<serde_json::Error> for RequestBuildError {
+    fn from(error: serde_json::Error) -> RequestBuildError {
+        RequestBuildError::Json { error }
+    }
+}
+
 struct LogRequestBuilder {
     default_api_key: Arc<str>,
-    encoding: JsonEncoding,
+    transformer: Transformer,
     compression: Compression,
 }
 
-impl RequestBuilder<(Option<Arc<str>>, Vec<Event>)> for LogRequestBuilder {
-    type Metadata = (Arc<str>, EventFinalizers);
-    type Events = Vec<Event>;
-    type Encoder = JsonEncoding;
-    type Payload = Bytes;
-    type Request = LogApiRequest;
-    type Error = RequestBuildError;
-
-    fn compression(&self) -> Compression {
-        self.compression
-    }
-
-    fn encoder(&self) -> &Self::Encoder {
-        &self.encoding
-    }
-
-    fn split_input(
+impl LogRequestBuilder {
+    fn build_request(
         &self,
-        input: (Option<Arc<str>>, Vec<Event>),
-    ) -> (Self::Metadata, RequestMetadataBuilder, Self::Events) {
-        let (api_key, mut events) = input;
-        let finalizers = events.take_finalizers();
-        let api_key = api_key.unwrap_or_else(|| Arc::clone(&self.default_api_key));
-        let builder = RequestMetadataBuilder::from_events(&events);
+        mut events: Vec<Event>,
+        api_key: Arc<str>,
+    ) -> Result<Vec<LogApiRequest>, RequestBuildError> {
+        // Transform events and pre-compute their estimated size.
+        let mut events_with_estimated_size: VecDeque<(Event, JsonSize)> = events
+            .iter_mut()
+            .map(|event| {
+                normalize_event(event);
+                self.transformer.transform(event);
+                let estimated_json_size = event.estimated_json_encoded_size_of();
+                (event.clone(), estimated_json_size)
+            })
+            .collect();
```

Comment:
Oops, I accidentally included an unnecessary clone here, which might be expensive.

```
    fn build_request(
        &self,
        events: Vec<Event>,
        api_key: Arc<str>,
    ) -> Result<Vec<LogApiRequest>, RequestBuildError> {
        // Transform events and pre-compute their estimated size.
        let mut events_with_estimated_size: VecDeque<(Event, JsonSize)> = events
            .into_iter()
            .map(|mut event| {
                normalize_event(&mut event);
                self.transformer.transform(&mut event);
                let estimated_json_size = event.estimated_json_encoded_size_of();
                (event, estimated_json_size)
            })
            .collect();
```

Reply:
That _should_ be cheap since the value is refcounted, but I was also curious where the previous perf bump went 😄

Reply:
Nice, the perf bump is back again 😄